### PR TITLE
Ensure Concurrency Keys are string-like and return a better error when they cannot be cast to a string

### LINKED ISSUE #784: Concurrency key doesn't handle Hash: TypeError (can't cast Hash)
As the title says. The job arguments are serialized by ActiveJob to something the database understands but the same is not true for the concurrency key.

```rb
class ExampleJob < ApplicationJob
  include GoodJob::ActiveJobExtensions::Concurrency
  good_job_control_concurrency_with(total_limit: 1, key: -> { arguments.first })

  def perform(arg)
  end
end

Job.new.perform({a: "test"})
# => All good
Job.perform_later({a: "test"})
# => Failed enqueuing Job to GoodJob(queue): TypeError (can't cast Hash)
```

Not sure if what I was trying to do should even be supported, but the error message I get as a result of doing this is not very clear and lead me to believe that the issues was with the parameters I was passing to the job.

Edit: Updated the issue description after figuring out what was actually going on.

### PR DESCRIPTION
**Pull request by Earlopain**
Closes #784 

If you pass something like a Hash or Array as the concurrency key, the resulting stack trace doesn't make it very clear what you did wrong:

```
   TypeError:
     can't cast Hash
   # ./app/models/good_job/lockable.rb:196:in `advisory_lock_key'
   # ./lib/good_job/active_job_extensions/concurrency.rb:51:in `block (2 levels) in <module:Concurrency>'
   # ./spec/lib/good_job/active_job_extensions/concurrency_spec.rb:157:in `block (4 levels) in <top (required)>'
   # ./spec/support/reset_rails_queue_adapter.rb:12:in `block (2 levels) in <top (required)>'
   # ./spec/support/reset_good_job.rb:19:in `block (2 levels) in <top (required)>'
   # ./spec/support/database_cleaner.rb:14:in `block (3 levels) in <top (required)>'
   # /usr/local/rvm/gems/default/gems/database_cleaner-core-2.0.1/lib/database_cleaner/strategy.rb:30:in `cleaning'
   # /usr/local/rvm/gems/default/gems/database_cleaner-core-2.0.1/lib/database_cleaner/cleaners.rb:34:in `block (2 levels) in cleaning'
   # /usr/local/rvm/gems/default/gems/database_cleaner-core-2.0.1/lib/database_cleaner/cleaners.rb:35:in `cleaning'
   # ./spec/support/database_cleaner.rb:13:in `block (2 levels) in <top (required)>'
```

This made me believe that I'm not allowed to pass a Hash to the job itself. Now the error is much clearer:

```
 ArgumentError:
       Concurrency key must be a String, was Hash
```

This is technically a breaking change. Some types like Number, Time and Symbol were being cast to a String by rails automatically, this now raises an error. The readme says that the concurrency key must be a String, but it would be best for this to go in the next major version if accepted.

### CODE REVIEW

File: lib/good_job/active_job_extensions/concurrency.rb [hunk 1]
```diff
@@ -117,11 +117,10 @@ def _good_job_concurrency_key
...
 
-        if key.respond_to? :call
-          instance_exec(&key)
-        else
-          key
-        end
+        key = instance_exec(&key) if key.respond_to? :call
+        raise ArgumentError, "Concurrency key must be a String, was #{key.class}" unless key.is_a?(String)
+
+        key
```
Comment (bensheldon) [inline lib/good_job/active_job_extensions/concurrency.rb:123]:
```suggestion
        key.to_s
```

**Review by bensheldon — COMMENTED** [review]
@Earlopain Thinking about this, it's too much of a breaking change: objects that previously could be coerced to a string will start raising exceptions.

Simply doing the coercion would avoid the issue you had of the value being cast, rather than coerced into a string (though still leaves open it not being understood that a complex object is being stringified)

**Comment by Earlopain** [comment]
> @Earlopain Thinking about this, it's too much of a breaking change: objects that previously could be coerced to a string will start raising exceptions.
> 
> Simply doing the coercion would avoid the issue you had of the value failing to cast (though still leaves open it not being understood that a complex object is being stringified)

I don't think doing `to_s` would be a good idea considering that you can pass model instances to jobs. It trades one issue for another:

```rb
class ExampleJob < ApplicationJob
  include GoodJob::ActiveJobExtensions::Concurrency
  good_job_control_concurrency_with(perform_limit: 1, key: -> { arguments.first })

  def perform(arg)
    sleep 60
  end
end

ExampleJob.perform_later(ExampleModel.first)
# key: "#<ExampleModel:0x00007f892b76bea0>"
ExampleJob.perform_later(ExampleModel.first)
# key: "#<ExampleModel:0x00007f892d66ee38>"
```
Both jobs would execute in parallel.

I agree on it being a breaking change. I have no issues with this being put on the backlog for v4, whenever that might be.

**Comment by bensheldon** [comment]
Hmmm. I don't want to _require_ it to be `is_a?(String)` because there are plenty of other classes that I think should be acceptable: String, Symbol, Numeric, Date/Time....

...though maybe that isn't as infinite as I feared and it would be fine to typecheck those.

I also wonder if it would be possible to catch the bad cast exception and raise an easier to understand error.

**Comment by Earlopain** [comment]
I like the approach of just catching the error. I changed the commit to do just that.

**Pull request merged**